Ethans gconf comments

ethanfrey opened the pull request:
Code speaks a thousand docs.

I liked this approach. But made some changes. Please look at the diff.

1. removed magic. pass in the package name/key rather than auto-querying it
2. enforce validation on parsing
3. abstracted out some basic InitConfig logic to make it easy to integrate in each package
4. Ensure tests in gconf and x/cash pass
5. changed genesis format to 
```json
"conf": {
  "cash": {..},
...
```

Please review the code and let me know what you think. If you are good, we can clean up and use this approach

husio commented on gconf/gconf.go:
```diff
@@ -46,17 +54,27 @@ type Unmarshaler interface {
-	if t.Kind() != reflect.Struct {
-		return ""
+	if confOptions[pkg] == nil {
+		return errors.Wrapf(errors.ErrInvalidInput, "No configuration for %s", pkg)
```
Please don't capitalize error strings https://github.com/golang/go/wiki/CodeReviewComments#error-strings

husio commented on gconf/gconf.go:
```diff
@@ -1,18 +1,22 @@
 
-func Save(db Store, src Marshaler) error {
-	key := []byte("configuration:" + pkgPath(src))
+// Save will validate code
```
Please write useful documentation :wink: :policeman:

husio commented on x/cash/codec.proto:
```diff
@@ -35,6 +35,7 @@ message FeeInfo {
 
 
 message Configuration {
+  // TODO: add schema uint32 here
```
This should land in https://github.com/iov-one/weave/issues/532 after merging. I think it's better to implement it before merging.

ethanfrey commented:
Yeah, I was a bit sloppy on the details. If Alex likes the idea as well, we can flesh it out. I would also add validation rules and see how to update schema/config before merging to master. 

But if we agree this is the direction we go, we can start working on the details. In office today or remote

ethanfrey commented:
@husio This branch should be good to go.

I added the generic handler with some basic tests and also fixed all tests in cmd/* and examples/mycoin.  Please run `make test` locally

If you like this, please merge it into your PR, rebase that PR on newest master, and then add any more changes you want to do to consider this finished. I am sure there is plenty left to improve, but the basic flow should work and captures all difficult logic in gconf.

ethanfrey commented:
Oh, and this will need a nicely updated CHANGELOG before merging to master :wink:
